### PR TITLE
Fix: Element dragging after ctrl + click on Mac [ED-1827]

### LINKED ISSUE #5228: Element dragging after right click on Mac ( when I didn't select anything in the right click menu )
I'm using Chrome on a Macbook Pro running El Capitan 
I have Elementor 2.1.5 and Elementor Pro 2.0.17. 

The issue started when Elementor implemented the right click UX.

IF I use right click and then see right click menu BUT don't click on any option in the right click menu, and then move mouse to another element / section, the element is stuck to my mouse and starts to move with it. ( the same way it does if I'm purposely trying to move an element )

I have to tap my mouse on the place where the element was for it to stop sticking to my mouse.

IF I wasn't paying attention or a client with less knowledge kept dragging the element around, then when the element was placed into the layout again it might be in the wrong spot!

image

I don't want to have to click on the layout to stop the element from being dragged around, I just want to right click, look at the right click menu, BUT be able to change my mind and move to a different element with nothing stuck to my mouse.

### PR DESCRIPTION
**Pull request by Nevoss**
fixes #5228

### CODE REVIEW

File: assets/dev/js/editor/component.js [hunk 1]
```diff
@@ -36,6 +37,10 @@ export default class Component extends ComponentBase {
 		return this.importCommands( internalCommands );
 	}
 
+	defaultUiStates() {
+		return this.importUiStates( uiStates );
+	}
```
Comment (StyleShit) [inline assets/dev/js/editor/component.js:42]: YAY 🚀

File: assets/dev/js/editor/ui-states/context-menu-mode.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+		return {
+			on: '',
+		};
+	}
+
+	onChange( oldValue, newValue ) {
+		this.toogleElementsSort( newValue );
+	}
+
+	toogleElementsSort( value ) {
```
Comment (StyleShit) [inline assets/dev/js/editor/ui-states/context-menu-mode.js:18]:
```suggestion
	toggleElementsSort( value ) {
```

Comment (StyleShit) [inline assets/dev/js/editor/ui-states/context-menu-mode.js:18]: Maybe change it so the function will receive a boolean and won't meed to know about the `on` option?

( Like any other generic toggle function which is boolean-based )

File: assets/dev/js/editor/ui-states/context-menu-mode.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import UiStateBase from 'elementor-api/core/states/ui-state-base';
+
+export class ContextMenuMode extends UiStateBase {
+	getId() {
+		return 'context-menu-mode';
+	}
+
+	getOptions() {
+		return {
+			on: '',
```
Comment (StyleShit) [inline assets/dev/js/editor/ui-states/context-menu-mode.js:10]: Just as a good practice, maybe extract it to a constant?

```JS
const MODE_ON = 'on';
```

And then use this constant in every place where you set the UI state.

File: core/common/assets/js/api/core/states/ui-state-base.js [hunk 4]
```diff
@@ -93,6 +96,16 @@ export default class UiStateBase {
 		};
 	}
 
+	/**
+	 * Triggered every UI state change.
+	 *
+	 * @param oldValue
+	 * @param newValue
```
Comment (StyleShit) [inline core/common/assets/js/api/core/states/ui-state-base.js:103]:
```suggestion
	 * Triggered on every UI state change.
	 *
	 * @param {string} oldValue
	 * @param {string} newValue
```

File: core/common/assets/js/api/core/states/ui-state-base.js [hunk 5]
```diff
@@ -32,6 +32,9 @@ export default class UiStateBase {
 		// Change the current state.
 		this.currentState = newValue;
 
+		// Trigger onChange method before option callback
```
Comment (StyleShit) [inline core/common/assets/js/api/core/states/ui-state-base.js:35]:
```suggestion
		// Trigger the `onChange` method before the option's callback.
```

File: core/common/assets/js/api/core/states/ui-state-base.js [hunk 6]
```diff
@@ -93,6 +96,16 @@ export default class UiStateBase {
...
 	}
 
+	/**
+	 * Triggered every UI state change.
+	 *
+	 * @param oldValue
+	 * @param newValue
+	 */
+	onChange( oldValue, newValue ) {
+		// Override this method if needed
```
Comment (StyleShit) [inline core/common/assets/js/api/core/states/ui-state-base.js:106]:
```suggestion
		// Override this method if needed.
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: core/common/assets/js/api/core/states/ui-state-base.js (hunk 4 above)
Comment (Nevoss) [reply core/common/assets/js/api/core/states/ui-state-base.js:103]: Done

File: assets/dev/js/editor/ui-states/context-menu-mode.js (hunk 2 above)
Comment (Nevoss) [reply assets/dev/js/editor/ui-states/context-menu-mode.js:18]: DOne

Comment (Nevoss) [reply assets/dev/js/editor/ui-states/context-menu-mode.js:18]: Done

File: assets/dev/js/editor/ui-states/context-menu-mode.js (hunk 3 above)
Comment (Nevoss) [reply assets/dev/js/editor/ui-states/context-menu-mode.js:10]: Done

File: core/common/assets/js/api/core/states/ui-state-base.js (hunk 5 above)
Comment (Nevoss) [reply core/common/assets/js/api/core/states/ui-state-base.js:35]: Done

File: core/common/assets/js/api/core/states/ui-state-base.js (hunk 6 above)
Comment (Nevoss) [reply core/common/assets/js/api/core/states/ui-state-base.js:106]: Done

Comment (Nevoss) [reply core/common/assets/js/api/core/states/ui-state-base.js]: Done

File: core/common/assets/js/api/core/states/ui-state-base.js (hunk 5 above)
Comment (Nevoss) [reply core/common/assets/js/api/core/states/ui-state-base.js]: Done

File: core/common/assets/js/api/core/states/ui-state-base.js (hunk 4 above)
Comment (Nevoss) [reply core/common/assets/js/api/core/states/ui-state-base.js]: Done

File: assets/dev/js/editor/ui-states/context-menu-mode.js (hunk 3 above)
Comment (Nevoss) [reply assets/dev/js/editor/ui-states/context-menu-mode.js]: Done

File: assets/dev/js/editor/ui-states/context-menu-mode.js (hunk 2 above)
Comment (Nevoss) [reply assets/dev/js/editor/ui-states/context-menu-mode.js]: Done

**Review by StyleShit — APPROVED** [review]

File: assets/dev/js/editor/ui-states/context-menu-mode.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+			[ this.constructor.MODE_ON ]: '',
+		};
+	}
+
+	onChange( oldValue, newValue ) {
+		this.toggleElementsSort( this.constructor.MODE_ON !== newValue );
+	}
+
+	toggleElementsSort( isActive ) {
+		elementor.getPreviewContainer().forEachChildrenRecursive( ( child ) => {
```
Comment (matipojo) [inline assets/dev/js/editor/ui-states/context-menu-mode.js:21]: Create a very large page and check the performance. (fast double context-click)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/elements/views/behaviors/context-menu.js [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import AddSectionBase	from 'elementor-views/add-section/base';
+import ContextMenuMode from '../../../ui-states/context-menu-mode';
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply assets/dev/js/editor/elements/views/behaviors/context-menu.js:2]: Done

**Review by matipojo — APPROVED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Please pull from develop

**Pull request merged**